### PR TITLE
[Attention] Enable dense and masked MHA for GLM-5

### PR DESCRIPTION
**Pull request by MatthewBonanni**
## Purpose

Depends on https://github.com/vllm-project/flash-attention/pull/179.

This PR enables the sparse-MLA masked-MHA prefill path for the GLM-5 256/256 QK/V head dimensions:

- Temporarily pins FlashAttention to `MatthewBonanni/flash-attention@8f844a0`, which contains head-dim-256 `mask_mod` support.
- Admits the GLM-5 dimensions `(64 heads, KV rank 512, QK 192+64, V 256)` to the existing masked-MHA route.
- Keeps the general FA4 head-dim-256 selector disabled. Sparse masked MHA uses contiguous projected K/V, while paged decoder attention still has unsupported non-page-aligned `max_seqlen_k` cases.
- Adds backend-specific GLM-5 routing thresholds derived from masked-MHA-vs-sparse-MQA benchmarks while leaving the existing DeepSeek thresholds unchanged.
- Extends the existing sparse-MLA reference test across DeepSeek 192/128 and GLM-5 256/256 dimensions, including cached context.

No other open PR was found for GLM-5 masked-MHA support.

### Routing policy and workspace guard

The threshold table records the benchmark-derived performance policy, including winning regions above 32K so it remains useful if the workspace limit changes. The existing workspace-capacity check is an independent final guard. With the current fixed 128 MiB workspace, a single pure-prefill request fits at exactly 32K and requests above 32K fall back to sparse MQA even when admitted by the performance policy. Batch size and cached context can make the workspace guard apply earlier.

The measured FlashMLA pure-prefill MHA-to-MQA crossovers were approximately 8.5K, 20.8K, 52.8K, and 116.8K tokens for TP1/2/4/8. The policy uses rounded conservative limits:

| TP | FlashMLA pure-prefill policy limit |
|---:|---:|
| 1 | 8K |
| 2 | 20K |
| 4 | 48K |
| 8 | 112K |

Cached-context routing uses a separate query/sequence-length table. TP1 stays on MQA when context is present; TP2/4/8 select masked MHA only inside benchmarked winning regions.

### FlashInfer routing

GLM-5 selects FlashInfer by default when the local head count is at most 16, so both TP4 and TP8 have benchmark-derived entries.

For TP4 pure prefill, the MQA/masked-MHA latency ratios were 4.70x at 4K, 2.09x at 16K, 1.19x at 32K, and 0.97x at 40K. The policy uses a conservative 36K limit. Its cached-context table uses the measured winning regions through q24K + 4K context.

For TP8, masked MHA remained faster throughout the valid pure-prefill range. The ratios were 8.61x at 4K, 4.39x at 16K, 2.60x at 32K, 1.81x at 48K, and 1.43x at 63K. The policy uses a 64K limit. Measurements continued through 64,512 query tokens; FlashInfer sparse MQA returned `CUDA_ERROR_INVALID_VALUE` at exactly 65,536 tokens, so that failed launch was excluded.

The fork URL and commit pin will be changed back to the vLLM FlashAttention repository after the dependency lands.

## Tests

- `.venv/bin/python -m pytest tests/v1/attention/test_sparse_mla_backends.py -q -k sparse_backend_prefill_correctness`
  - 8 passed.
- `.venv/bin/python -m pytest tests/v1/attention/test_mla_backends.py -q -k 'glm5 or dimension_specific or unknown_backend'`
  - 32 passed.
- `.venv/bin/python -m pytest tests/v1/attention/test_sparse_mla_backends.py -q -k masked_mha_workspace`
  - 7 passed, including explicit checks that long policy entries are rejected by the current workspace guard.
- `.venv/bin/python -m pytest tests/models/multimodal/pooling/test_colpali.py::test_colpali_v2_multimodal_text_query_image_docs -v`
  - 1 passed.
- `.venv/bin/pre-commit run --files benchmarks/attention_benchmarks/benchmark.py benchmarks/attention_benchmarks/configs/mla_sparse_masked_mha_vs_mqa_glm5.yaml tests/v1/attention/test_mla_backends.py tests/v1/attention/test_sparse_mla_backends.py vllm/model_executor/layers/attention/mla_attention.py cmake/external_projects/vllm_flash_attn.cmake`
  - Passed.
- Consolidated-config smoke test with `q4096`, FlashInfer, four local-head counts, and both MHA/MQA variants.
  - 8 benchmark cases completed.

- `.venv/bin/python -m pytest tests/model_executor/layers/test_mla_short_prefill_indexer.py -q`
  - 9 passed, including BF16 and FP8 query construction for selected dense MHA.
- `.venv/bin/python -m pytest tests/v1/attention/test_mla_backends.py -q -k masked_mha`
  - 32 passed.
- `.venv/bin/pre-commit run --files vllm/models/deepseek_v32/attention.py tests/model_executor/layers/test_mla_short_prefill_indexer.py`
  - Passed.

## Benchmark commands

The single config sweeps 64/32/16/8 local query heads, corresponding to GLM-5 TP1/2/4/8.

```bash
results=benchmark_outputs/glm5_hd256_masked_mha_vs_mqa
config=benchmarks/attention_benchmarks/configs/mla_sparse_masked_mha_vs_mqa_glm5.yaml
mkdir -p "${results}"

.venv/bin/python benchmarks/attention_benchmarks/benchmark.py \
  --config "${config}" \
  --output-csv "${results}/all_backends.csv"

.venv/bin/python benchmarks/attention_benchmarks/benchmark.py \
  --config "${config}" \
  --backends FLASHMLA_SPARSE \
  --batch-specs q32768 q40960 q49152 q65536 q98304 q131072 \
  --output-csv "${results}/flashmla_pure_prefill_extended.csv"

.venv/bin/python benchmarks/attention_benchmarks/benchmark.py \
  --config "${config}" \
  --backends FLASHINFER_MLA_SPARSE \
  --batch-specs q32768 q40960 q49152 q57344 q61440 q64512 \
  --output-csv "${results}/flashinfer_pure_prefill_extended.csv"

context_specs=()
for q in 4096 8192 16384 24576 32768 40960 49152 57344 64512; do
  for context in 1024 2048 4096 8192 16384; do
    context_specs+=("q${q}s$((q + context))")
  done
done

.venv/bin/python benchmarks/attention_benchmarks/benchmark.py \
  --config "${config}" \
  --batch-specs "${context_specs[@]}" \
  --output-csv "${results}/cached_context.csv"
```

speedup_curve_glm5_hd256
speedup_context_heatmap_glm5_hd256

### End-to-end GLM-5.2 benchmark

`nvidia/GLM-5.2-NVFP4`, TP8 on 8x NVIDIA B200, ISL 8192, OSL 1, concurrency 1, and 16 requests:

| Metric | Main | Masked MHA PR | Change |
|---|---:|---:|---:|
| Request throughput | 2.12 req/s | 3.90 req/s | **+83.8%** |
| Total token throughput | 17,383 tok/s | 31,949 tok/s | **+83.8%** |
| Mean TTFT | 471.16 ms | 256.27 ms | **-45.6%** |
| P99 TTFT | 477.76 ms | 262.24 ms | **-45.1%** |
| Mean TPOT | 0.00 ms | 0.00 ms | N/A |
| P99 TPOT | 0.00 ms | 0.00 ms | N/A |
| Mean ITL | 0.00 ms | 0.00 ms | N/A |
| P99 ITL | 0.00 ms | 0.00 ms | N/A |
| Successful requests | 16/16 | 16/16 | — |

TPOT and ITL are zero because OSL is 1, so there are no subsequent decode tokens.

### Generation regression coverage

Generation was exercised around the short-prefill transition that previously exposed dense-MHA corruption, as well as with long prefill and cached context:

| Configuration | Prompt lengths | Result |
|---|---:|---|
| TP8, BF16 KV, piecewise CUDA graphs | 97-1001 tokens (19 values, including 127/129) | 19/19 returned the expected content |
| TP8, BF16 KV, pure prefill | 8192 tokens | Correct retrieval |
| TP8, BF16 KV, prefix cache | 4000 cached + approximately 4100 query tokens | Correct retrieval at 8100 total tokens |
| Prefill context parallel size 4, FP8 KV | 127, 129, 202, and 1001 tokens | 4/4 returned the expected content |

The 129- and 202-token reproductions no longer produce the repetitive numeric output observed before the CUDA-graph routing fix.

## Model evaluation

The sparse-MLA tests compare kernel output against an SDPA reference for both head-dimension configurations. The GLM-5 pure-prefill and cached-context masked-MHA cases passed. The ColPali multimodal scoring smoke test also passed.

## AI assistance disclosure

OpenAI Codex assisted with implementation, testing, benchmarking, and drafting this PR. The submitter reviewed every changed line and the reported results.

### CODE REVIEW

**Comment by simon-veitner-redhat** [comment]
There is a problem with the selection logic. Report created with Codex:

---

Production allocates a fixed 128 MiB workspace:

```python
# vllm/model_executor/layers/attention/sparse_mla_attention.py
GLOBAL_TOPK_MASK_MAX_BYTES = 128 * 1024 * 1024

self.topk_mask_workspace = torch.zeros(
    GLOBAL_TOPK_MASK_MAX_BYTES // torch.int32.itemsize,
    dtype=torch.int32,
    device=device,
)
```

The actual route requires both policy admission and sufficient workspace:

```python
use_masked_mha = (
    ...
    and _use_masked_mha(...)
    and self.impl.masked_mha_workspace_fits(prefill)
)
```

The workspace formula is derived from `_topk_mask_shape()` and `_masked_mha_workspace_fits()` in `sparse_mla_attention.py:51-77`:

```python
tile_m = 128 if max_query_len <= 128 else 256
padded_q_len = triton.cdiv(max_query_len, tile_m) * tile_m
num_words = triton.cdiv(max_key_len, 32) + int(reserve_key_starts_word)
num_words = triton.cdiv(num_words, 4) * 4
return batch_size, padded_q_len, num_words

max_key_len = max(max_query_len, max_context_chunk_seq_len)
needed = math.prod(
    _topk_mask_shape(batch_size, max_query_len, max_key_len)
)
```

The mandatory capacity check uses the default `reserve_key_starts_word=False`.

For long requests, this shape is:

```text
(B, 256 * ceil(Q / 256), 4 * ceil(K / 128))
K = max(Q, max_context_chunk_seq_len)
```

Derivation:

- Long requests have `Q > 128`, so `tile_m = 256`; `cdiv(Q, 256) * 256` is `256 * ceil(Q / 256)`.
- With `reserve_key_starts_word=False`, the first `cdiv` needs `ceil(K / 32)` int32 mask words. The second `cdiv` rounds that count to a multiple of four: `4 * ceil(ceil(K / 32) / 4) = 4 * ceil(K / 128)`.
- The first dimension remains `B`, and `_masked_mha_workspace_fits()` defines `K = max(Q, max_context_chunk_seq_len)`.

`needed` counts `int32` elements, so multiplying the shape by four bytes gives the required workspace. For one aligned pure-prefill request, `B = 1` and `K = Q`, which simplifies to `Q^2 / 8` bytes.

| Pure-prefill length | Required workspace | Fits 128 MiB? |
|---:|---:|:---:|
| 32K | 128 MiB | yes, exactly |
| 48K | 288 MiB | no |
| 64K | 512 MiB | no |
| 112K | 1,568 MiB | no |

The reachable portion of a single-request route therefore ends at 32,768 tokens. Batch size multiplies the requirement, so multi-request batches can hit the limit earlier.

---

**Comment by simon-veitner-redhat** [comment]
Another problem with selection logic:

---
GLM-5's `GlmMoeDsaForCausalLM` reuses the DeepSeek-V3.2 DSA implementation (`vllm/models/deepseek_v32/__init__.py:17-22`). Its attention class computes local heads here:

```python
# vllm/models/deepseek_v32/attention.py:141-145
num_heads = config.num_attention_heads
tp_size = get_tensor_model_parallel_world_size()
num_local_heads = num_heads // tp_size  # 64 // 4 = 16
```

On SM10, 16 or fewer heads prioritizes FlashInfer (`vllm/platforms/cuda.py:105-116`):

```python
if num_heads is not None and num_heads <= 16:
    sparse_backends = [
        FLASHINFER_MLA_SPARSE,
        FLASHMLA_SPARSE,
    ]
```

The new table contains FlashMLA TP4 but FlashInfer TP8 only (`mla_attention.py:1663-1702`):

```python
(256, 256): {
    "FLASHMLA_SPARSE": {4: (((48 * 1024, 0),), ...)},
    "FLASHINFER_MLA_SPARSE": {8: (((64 * 1024, 0),), ...)},
}
```

Routing uses the selected backend name:

```python
thresholds = table.get(dimensions, {}).get(backend_name, {}).get(tp_size)
if thresholds is None:
    return False
```

Therefore default TP4 does this:

```text
64 heads / TP4 = 16 local heads
-> select FLASHINFER_MLA_SPARSE
-> look for FlashInfer TP4 thresholds
-> no entry
-> use FlashInfer sparse MQA, not FA4 masked MHA
```
---

File: benchmarks/attention_benchmarks/configs/mla_sparse_masked_mha_vs_mqa_glm5_tp8.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  v_head_dim: 256
+  block_size: 128
+  max_model_len: 131072
+batch_specs: ["q256s2048", "q512s2048", "q1024s2048", "q1536s2048", "q2048", "q256s4096", "q512s4096", "q1024s4096", "q1536s4096", "q2048s4096", "q4096", "q256s8192", "q512s8192", "q1024s8192", "q1536s8192", "q2048s8192", "q4096s8192", "q8192", "q256s16384", "q512s16384", "q1024s16384", "q1536s16384", "q2048s16384", "q4096s16384", "q8192s16384", "q16384", "q256s32768", "q512s32768", "q1024s32768", "q1536s32768", "q2048s32768", "q4096s32768", "q8192s32768", "q16384s32768", "q32768"]
+backends: ["FLASHMLA_SPARSE"]
+device: "cuda:0"
+profile_memory: false
+sparse_mla_masked_mha_max_seq_len: 131072
+sparse_mla_mha_variants: ["masked_mha", "mqa"]
+sparse_mla_topk_pattern: "random"
```
Comment (simon-veitner-redhat) [inline benchmarks/attention_benchmarks/configs/mla_sparse_masked_mha_vs_mqa_glm5_tp8.yaml]: can we summarize into one big config?

File: benchmarks/attention_benchmarks/plot_mla_sparse_mha_vs_mqa.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        "Speedup = sparse MQA latency / masked MHA latency\n"
+        ">1 favors masked MHA; black contour = 1.0"
+    )
+    figure.suptitle("Sparse MLA routing sweep on B200", fontsize=17)
+    output = args.output or args.input_dir / "speedup_heatmap_tp1_tp2_tp4_tp8.png"
+    figure.savefig(output, dpi=180)
+
+
+if __name__ == "__main__":
+    main()
```
Comment (simon-veitner-redhat) [inline benchmarks/attention_benchmarks/plot_mla_sparse_mha_vs_mqa.py]: should we keep this script here?

File: benchmarks/attention_benchmarks/plot_mla_sparse_mha_vs_mqa.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        axis.set_xlabel("Query length")
+        axis.set_facecolor("#eee")
+    axes[0].set_ylabel("Total sequence length")
+    assert image is not None
+    colorbar = figure.colorbar(image, ax=axes, shrink=0.93, pad=0.012)
+    colorbar.set_label(
+        "Speedup = sparse MQA latency / masked MHA latency\n"
+        ">1 favors masked MHA; black contour = 1.0"
+    )
+    figure.suptitle("Sparse MLA routing sweep on B200", fontsize=17)
```
Comment (simon-veitner-redhat) [inline benchmarks/attention_benchmarks/plot_mla_sparse_mha_vs_mqa.py]: specific to B200

File: tests/v1/attention/test_mla_backends.py [hunk 4]
```diff
@@ -68,6 +69,115 @@
...
+        backend_name="FLASHMLA_SPARSE",
+        tensor_parallel_size=1,
+        qk_head_dim=128,
+        v_head_dim=128,
+        query_len=1536,
+        seq_len=2048,
+        has_context=True,
+    )
+
+
```
Comment (simon-veitner-redhat) [inline tests/v1/attention/test_mla_backends.py:189]: tests not sufficient (workspace issue described above)

**Comment by MatthewBonanni** [comment]
Addressed all comments. Regarding the seq_len thresholds, I made the table intentionally document crossovers greater than 32k. `masked_mha_workspace_fits()` prevents the 128 MiB workspace from being violated. I added a test to verify this

**Comment by simon-veitner-redhat** [comment]
LGTM

**Review by LucasWilkinson — APPROVED** [review]

**Comment by simon-veitner-redhat** [comment]
I think this needs further work so we can leverage the improved kernel performance for GLM-5.2 before we merge.

---

PR [#52512](https://github.com/vllm-project/vllm/pull/52512) disabled dense MHA after this sequence caused repeated tokens:

1. Short-prefill metadata selected dense MHA.
2. The indexer therefore skipped top-k scoring and left the top-k buffer filled with `-1`.
3. `DeepseekV32Attention.forward` did not run dense MHA. It always ran sparse MQA.
4. Sparse MQA consumed the invalid top-k buffer.

The fix is still present in this PR:

[`vllm/models/deepseek_v32/attention.py:126`](https://github.com/MatthewBonanni/vllm/blob/d0aa1d9bf7b066b96e81e71aa262838663cadf71/vllm/models/deepseek_v32/attention.py#L126)

```python
supports_dense_mha_prefill = False
```

---

https://github.com/MatthewBonanni/vllm/blob/d0aa1d9bf7b066b96e81e71aa262838663cadf71/vllm/models/deepseek_v32/attention.py#L126

```python
class DeepseekV32Attention(MLAAttention):
    indexer: "DeepseekV32Indexer | None"
    indexer_cls: "type[DeepseekV32Indexer]" = DeepseekV32Indexer

    supports_dense_mha_prefill = False
```

From this, it follows that in https://github.com/MatthewBonanni/vllm/blob/d0aa1d9bf7b066b96e81e71aa262838663cadf71/vllm/model_executor/layers/attention/mla_attention.py#L567-L575:

```python
if self.impl.is_sparse and not (
    self.impl.supports_dense_mha_prefill and self.supports_dense_mha_prefill
):
    logger.warning_once(
        "Sparse MLA layer has no dense-MHA prefill path; using the top-k "
        "MQA path only."
    )
    self.prefill_backend = None
```

and:

```python
use_masked_mha = (
    self.prefill_backend is not None
    and self.impl.masked_mha_available
    ...
)
```

Furthermore, in `DeepseekV32Attention.forward` from above:

```python
attn_out, lse = self.impl.forward_mqa(  # type: ignore[attr-defined]
    mqa_q_arg, kv_cache, attn_metadata, self
)
```
